### PR TITLE
Add powershellcontext

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
So.............. here it is.

Most is copy pasted of course.

I've taken:
PowerShellEditorServices/Console
PowerShellEditorServices/Extensions
PowerShellEditorServices/Session

and moved them under `PowerShellEditorServices.Engine/Services/PowerShellContext/` respectively...

and renamed `PowerShellContext.cs` to `PowerShellContextService.cs` to align with the service model.

This gives you the integrated console we know and love

Take a look at the changes to EditorServicesHost and the Configuration handler for things that tied it all together.

TODO:
- [x] ~set `$psEditor`~

### CODE REVIEW

**Comment by TylerLeonhardt** [comment]
I've decided that psEditor will be in another PR as it involves even more new files to set up the messages from the server to the client

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Components/IComponentRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        /// </param>
+        /// <param name="componentInstance">
+        /// The out parameter in which the found instance will be stored.
+        /// </param>
+        /// <returns>
+        /// True if a registered instance was found, false otherwise.
+        /// </returns>
+        bool TryGet(Type componentType, out object componentInstance);
+    }
+}
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Components/IComponentRegistry.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Components/IComponentRegistryExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            if (componentRegistry.TryGet(typeof(TComponent), out componentObject))
+            {
+                componentInstance = componentObject as TComponent;
+                return componentInstance != null;
+            }
+
+            return false;
+        }
+    }
+}
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Components/IComponentRegistryExtensions.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Console/ChoicePromptHandler.cs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+            return
+                choiceArray != null
+                ? choiceArray.DefaultIfEmpty(-1).First()
+                : -1;
+        }
+
+        #endregion
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Console/ChoicePromptHandler.cs]: One newline too many?

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/EditorCommandAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+        /// <summary>
+        /// Gets or sets a value indicating whether this command's output
+        /// should be suppressed.
+        /// </summary>
+        public bool SuppressOutput { get; set; }
+
+        #endregion
+    }
+}
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/EditorCommandAttribute.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/EditorContext.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        {
+            this.editorOperations
+                .SetSelectionAsync(selectionRange)
+                .Wait();
+        }
+
+        #endregion
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/EditorContext.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/EditorObject.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        /// at the time this method is invoked.
+        /// </summary>
+        /// <returns>A instance of the EditorContext class.</returns>
+        public EditorContext GetEditorContext()
+        {
+            return this.editorOperations.GetEditorContextAsync().Result;
+        }
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/EditorObject.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/EditorWorkspace.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        /// <param name="preview">Determines wether the file is opened as a preview or as a durable editor.</param>
+        public void OpenFile(string filePath, bool preview)
+        {
+            this.editorOperations.OpenFileAsync(filePath, preview).Wait();
+        }
+
+        #endregion
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/EditorWorkspace.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/ExtensionService.cs [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+        private void OnCommandRemoved(EditorCommand command)
+        {
+            this.CommandRemoved?.Invoke(this, command);
+        }
+
+        #endregion
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/ExtensionService.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/FileContext.cs [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+                throw new IOException(String.Format("The file '{0}' already exists", absolutePath));
+            }
+
+            this.editorOperations.SaveFileAsync(this.scriptFile.FilePath, newFilePath);
+        }
+
+        #endregion
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/FileContext.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/IEditorOperations.cs [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        /// <summary>
+        /// Sets the status bar message in the editor UI (if applicable).
+        /// </summary>
+        /// <param name="message">The message to be shown.</param>
+        /// <param name="timeout">If non-null, a timeout in milliseconds for how long the message should remain visible.</param>
+        /// <returns>A Task that can be tracked for completion.</returns>
+        Task SetStatusBarMessageAsync(string message, int? timeout);
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Extensions/IEditorOperations.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/Host/IHostInput.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        /// Stops the host's interactive command loop.
+        /// </summary>
+        void StopCommandLoop();
+
+        /// <summary>
+        /// Cancels the currently executing command or prompt.
+        /// </summary>
+        void SendControlC();
+    }
+}
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/Host/IHostInput.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/OutputWrittenEventArgs.cs [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            this.OutputText = outputText;
+            this.IncludeNewLine = includeNewLine;
+            this.OutputType = outputType;
+            this.ForegroundColor = foregroundColor;
+            this.BackgroundColor = backgroundColor;
+        }
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/OutputWrittenEventArgs.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/PowerShell5Operations.cs [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+            {
+                host.ExitNestedPrompt();
+            }
+            catch (FlowControlException)
+            {
+            }
+        }
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/PowerShell5Operations.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/PowerShellContextState.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        Aborting,
+
+        /// <summary>
+        /// Indicates that the session is already disposed and cannot
+        /// accept further execution requests.
+        /// </summary>
+        Disposed
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/PowerShellContextState.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/PowerShellExecutionResult.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        /// </summary>
+        Stopped,
+
+        /// <summary>
+        /// Indicates that execution completed successfully.
+        /// </summary>
+        Completed
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/PowerShellExecutionResult.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/ProgressDetails.cs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+            //progressRecord.
+
+            return new ProgressDetails
+            {
+                PercentComplete = progressRecord.PercentComplete
+            };
+        }
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/ProgressDetails.cs]: Newline

File: src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/SessionStateChangedEventArgs.cs [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+            PowerShellExecutionResult executionResult,
+            Exception errorException)
+        {
+            this.NewSessionState = newSessionState;
+            this.ExecutionResult = executionResult;
+            this.ErrorException = errorException;
+        }
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Services/PowerShellContext/Session/SessionStateChangedEventArgs.cs]: Newline

File: src/PowerShellEditorServices.Engine/Utility/AsyncLock.cs [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+            public void Dispose()
+            {
+                this.lockToRelease.lockSemaphore.Release();
+            }
+        }
+
+        #endregion
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Utility/AsyncLock.cs]: Newline

File: src/PowerShellEditorServices.Engine/Utility/AsyncQueue.cs [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+                }
+            }
+
+            return requestTask.Task.GetAwaiter().GetResult();
+        }
+
+        #endregion
+    }
+}
+
```
Comment (rjmholt) [inline src/PowerShellEditorServices.Engine/Utility/AsyncQueue.cs]: Newline

**Comment by TylerLeonhardt** [comment]
Not going to deal with Codacy issues for now since most of this was copy paste

**Pull request merged**